---
title: "IGV: programmatic access to the igv browser-based genome track viewer"
author: "Paul Shannon"
package: IGV
date: "`r Sys.Date()`"
output:
   BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{"IGV: programmatic access to the igv browser-based genome track viewer"}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
## Overview

The IGV package provides easy access to the web-based javascript ["igv.js"](https://github.com/igvteam/igv.js) in order to create and display genome tracks in a richly interactive web browser visual interface. 

We introduce IGV via a case study of genetic variants associated with Alzheimer's Disease as described in the 2014 Nature Genetics paper "Meta-analysis of 74,046 individuals identifies 11 new susceptibility loci for Alzheimerâ€™s disease"^[Lambert, Jean-Charles, et al. "Meta-analysis of 74,046 individuals identifies 11 new susceptibility loci for Alzheimer's disease." Nature genetics 45.12 (2013): 1452.]  We focus on a 1 megabase region surrounding the MEF2C gene, in which contaions 208 variants whose presence is associated with the expression of MEF2C in 17,008 Alzheimer's disease cases and 37,154 controls.  These variants do no occur in coding regions.  Some possible insight into their function -- of a highly speculative and preliminary sort - is provided by this case study.  We pursue the question:  could these non-coding variants after MEF2C gene regulation by disrupting transcription factor binding sites?

We will proceed as follows:

 - load the IGV package, connect to the igv.js genome browser
 - set the reference genome (hg19)
 - zoom into the MEF2C region 
 - expand to display the entire 1MB region of interest
 - create our first track - the 208 SNPs - using the most primitive approach, that is, with a base R data.frame
 - use Bioconductor's AnnotationHub to retrieve three methylation and five DNase hypersensitivity datasets: these two phenomenon are associated with open chromatin and transcription factor binding sites.  With each of these datasets we create an IGV track (of an approriate type - IGV's Track class has a variety of subclasses depending on data type and data location)
 - for good measure we conclude by loading a different variant dataset, in VCF format, from another Alzheimer's [study](http://adni.loni.usc.edu/data-samples/genetic-data/wgs/) of 818 participants.
 
 ## setup IGV
 
 ```{r startup, echo=FALSE}
library(IGV)
library(VariantAnnotation)
library(AnnotationHub)

igv <- IGV()
setBrowserWindowTitle(igv, "MEF2C")
setGenome(igv, "hg19")
Sys.sleep(5)   # wait a few seconds before zooming into MEF2C
showGenomicRegion(igv, "MEF2C")
```

## Create a GRanges object specifying our region of interest.

```{r roi, echo=TRUE}
chrom <- "chr5"
shoulder <- 50000
start.loc <- 88013975 - shoulder
end.loc   <- 88199922 + shoulder
mef2c.region <- GRanges(seqnames=chrom, IRanges(start=start.loc, end=end.loc))
showGenomicRegion(igv, list(chrom=chrom, start=start.loc, end=end.loc))
```

## Read a tab-delimited file, display as a "primitive" data.frame Track

These next few steps expose the nuts and bolts of creating and displaying tracks from tabular data.  None of the power and convenience of the Biocondcutor AnnotationHub, of rtracklayer and GenomicRanges is used here (that comes next).  

Add IGAP gwas twice, first as an annotation track, so that names are avaialble by clicking,
then again as a quantitative track, a -log10(pValue) score, conveying strength of association
of that snp with the expression of the target gene, MEF2C


```{r data.frame.track, echo=TRUE}
load("~/s/work/priceLab/AD/tbl.gwas.level_1.RData")
tbl.mef2cGWAS.variants <- subset(tbl.gwas, CHR=="chr5" & oldPos > start.loc & oldPos < end.loc)
tbl.mef2cGWAS.variants.bed <- tbl.mef2cGWAS.variants[, c("CHR", "oldPos", "oldPos", "SNP", "P")]
tbl.mef2cGWAS.variants.bed$P <- -log10(tbl.mef2cGWAS.variants.bed$P)
colnames(tbl.mef2cGWAS.variants.bed) <- c("chrom", "start", "end", "name", "score")
track.gwas <- DataFrameAnnotationTrack("IGAP.gwas", tbl.mef2cGWAS.variants.bed, trackHeight=20, color="darkBlue")
displayTrack(igv, track.gwas)

tbl.mef2cGWAS.variants.bedGraph <-  tbl.mef2cGWAS.variants.bed[, -4]
track.gwas.numeric <- DataFrameQuantitativeTrack("IGAP.gwas.scored",tbl.mef2cGWAS.variants.bedGraph)
displayTrack(igv, track.gwas.numeric)

```

