<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Explore VCF variants, GWAS snps, promoters and histone marks around the  MEF2C gene in Alzheimers Disease • igvR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Explore VCF variants, GWAS snps, promoters and histone marks around the  MEF2C gene in Alzheimers Disease">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">igvR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.45</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/alzheimersVariantsNearMEF2C.html">Explore VCF variants, GWAS snps, promoters and histone marks around the  MEF2C gene in Alzheimers Disease</a>
    </li>
    <li>
      <a href="../articles/basicIntro.html">Introduction: a simple demo</a>
    </li>
    <li>
      <a href="../articles/ctcfChipSeq.html">Explore ChIP-seq alignments from a bam file, MACS2 narrowPeaks, conservation, H3K4me3 methylation and motif matching</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Explore VCF variants, GWAS snps, promoters and histone marks around the MEF2C gene in Alzheimers Disease</h1>
                        <h4 class="author">Paul Shannon</h4>
            
            <h4 class="date">2019-10-23</h4>
      
      
      <div class="hidden name"><code>alzheimersVariantsNearMEF2C.Rmd</code></div>

    </div>

    
    
<style>
.main-container { width: 1200px; max-width:2800px;}
</style>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>The igvR package provides easy programmic access in R to the web-based javascript library <a href="https://github.com/igvteam/igv.js">igv.js</a> in order to create and display genome tracks in its richly interactive web browser visual interface. I am grateful to Jim Robinson, Douglass Turner and colleagues for their fine work.</p>
<p>We introduce igvR via a case study of genetic variants associated with Alzheimer’s Disease as described in the 2014 Nature Genetics paper</p>
<p>      <em>Meta-analysis of 74,046 individuals identifies 11 new susceptibility loci for Alzheimer’s disease</em><sup>1</sup></p>
<p>Theses 11 loci bring the total susceptibility loci count to 20, of which MEF2C is one. We focus on a 1 megabase region surrounding the MEF2C gene which contains 208 variants whose presence is associated with the expression of MEF2C in 17,008 Alzheimer’s disease cases and 37,154 controls. Interestingly, these variants do not occur in coding regions of the gene. Some possible insight into their function – of a highly speculative and preliminary sort - is provided by this case study. We pursue the question: could these non-coding variants alter MEF2C gene regulation by disrupting transcription factor binding sites?</p>
<p>We will proceed as follows:</p>
<ul>
<li><p>load the igvR package</p></li>
<li><p>connect to the igv.js genome browser</p></li>
<li><p>set the reference genome (hg19)</p></li>
<li><p>zoom into the MEF2C region</p></li>
<li><p>expand to display the entire 1MB region of interest</p></li>
<li><p>create our first track - the 208 SNPs - using the most primitive approach: from a base R data.frame</p></li>
<li><p>employ Bioconductor’s AnnotationHub to retrieve three methylation datasets - which are associated with open chromatin and transcription factor binding sites. From each of these datasets we create an igvR track of an approriate type - igvR’s Track class has a variety of subclasses depending on data type and data source location. Annotation, Quantitative and Variant tracks are supported here in this first (1.0) version of igvR.</p></li>
<li><p>for good measure we conclude by loading a very small slice of a different variant dataset, in VCF format, from another Alzheimer’s <a href="http://adni.loni.usc.edu/data-samples/genetic-data/wgs/">study</a> of 818 participants. VCF files typically contain fine-grained variant data, as is seen in the igv.js popup dialog.</p></li>
</ul>
<p>The result from running the code below:</p>
<p><img src="igvR-mef2c-Alzheimers-demo.png" width="730"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(igvR)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(MotifDb)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BSgenome.Hsapiens.UCSC.hg19)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(VariantAnnotation)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(AnnotationHub)</code></pre></div>
<p>Create the igvR instance, with all default parameters (portRange, quiet, title). Javascript and HTML is loaded into your browser, igv.js is initialized, a websocket connection between your R process and that web page is constructed, over which subsequent commands and data will travel.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">igv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/igvR-class.html">igvR</a></span>()
<span class="kw">setBrowserWindowTitle</span>(igv, <span class="st">"MEF2C"</span>)
<span class="kw"><a href="../reference/setGenome.html">setGenome</a></span>(igv, <span class="st">"hg19"</span>)</code></pre></div>
<p>Display MEF2C</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/showGenomicRegion.html">showGenomicRegion</a></span>(igv, <span class="st">"MEF2C"</span>)</code></pre></div>
</div>
<div id="create-a-granges-object-specifying-our-region-of-interest-285-kb-centered-on-the-mef2c-gene" class="section level1">
<h1 class="hasAnchor">
<a href="#create-a-granges-object-specifying-our-region-of-interest-285-kb-centered-on-the-mef2c-gene" class="anchor"></a>Create a GRanges object specifying our region of interest, 285 kb centered on the MEF2C gene</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chrom &lt;-<span class="st"> "chr5"</span>
shoulder &lt;-<span class="st"> </span><span class="dv">50000</span>
start.loc &lt;-<span class="st"> </span><span class="dv">88013975</span> <span class="op">-</span><span class="st"> </span>shoulder
end.loc   &lt;-<span class="st"> </span><span class="dv">88199922</span> <span class="op">+</span><span class="st"> </span>shoulder
mef2c.region &lt;-<span class="st"> </span><span class="kw">GRanges</span>(<span class="dt">seqnames=</span>chrom, <span class="kw">IRanges</span>(<span class="dt">start=</span>start.loc, <span class="dt">end=</span>end.loc))
<span class="kw"><a href="../reference/showGenomicRegion.html">showGenomicRegion</a></span>(igv, <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">chrom=</span>chrom, <span class="dt">start=</span>start.loc, <span class="dt">end=</span>end.loc))</code></pre></div>
<p>These next few steps expose the nuts and bolts of creating and displaying tracks from tabular data. None of the power and convenience of the Biocondcutor AnnotationHub, of rtracklayer and GenomicRanges is used here - that comes later.</p>
<p>Add the IGAP GWAS track twice, first as an annotation track, so that names are avaialble by clicking, then again as a quantitative track, using a -log10(pValue) score, conveying strength of association of each SNP with the expression of the target gene, MEF2C.</p>
<p><a href="http://consortiapedia.fastercures.org/consortia/igap/">IGAP</a>, the International Genomics of Alzheimer’s Project, provide these data. See <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3896259/">Meta-analysis of 74,046 individuals identifies 11 new susceptibility loci for Alzheimer’s disease</a>, Nature Genetics, Dec 2013.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="dt">package=</span><span class="st">"igvR"</span>, <span class="st">"extdata"</span>, <span class="st">"tbl.mef2cGWAS.variants.RData"</span>))
tbl.mef2cGWAS.variants.bed &lt;-<span class="st"> </span>tbl.mef2cGWAS.variants[, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CHR"</span>, <span class="st">"oldPos"</span>, <span class="st">"oldPos"</span>, <span class="st">"SNP"</span>, <span class="st">"P"</span>)]
tbl.mef2cGWAS.variants.bed<span class="op">$</span>P &lt;-<span class="st"> </span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(tbl.mef2cGWAS.variants.bed<span class="op">$</span>P)
<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(tbl.mef2cGWAS.variants.bed) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chrom"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"name"</span>, <span class="st">"score"</span>)
track.gwas &lt;-<span class="st"> </span><span class="kw"><a href="../reference/DataFrameAnnotationTrack-class.html">DataFrameAnnotationTrack</a></span>(<span class="st">"IGAP.gwas"</span>, tbl.mef2cGWAS.variants.bed, <span class="dt">trackHeight=</span><span class="dv">20</span>, <span class="dt">color=</span><span class="st">"darkBlue"</span>)
<span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track.gwas)

tbl.mef2cGWAS.variants.bedGraph &lt;-<span class="st">  </span>tbl.mef2cGWAS.variants.bed[, <span class="op">-</span><span class="dv">4</span>]
track.gwas.numeric &lt;-<span class="st"> </span><span class="kw"><a href="../reference/DataFrameQuantitativeTrack-class.html">DataFrameQuantitativeTrack</a></span>(<span class="st">"IGAP.gwas.scored"</span>, tbl.mef2cGWAS.variants.bedGraph, <span class="dt">autoscale=</span><span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track.gwas.numeric)</code></pre></div>
</div>
<div id="use-the-annotationhub-to-create-a-promoter-track" class="section level1">
<h1 class="hasAnchor">
<a href="#use-the-annotationhub-to-create-a-promoter-track" class="anchor"></a>Use the AnnotationHub to create a promoter track</h1>
<p>A location-only calculation is used to identify gene promoters, rather than more nuanced experimental data, of which we will seem more later in this vignette. We use default, conservative parameters (2kb upstream of TSS, 200bp downstream).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ah &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">AnnotationHub</a></span>()
ah.human &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(ah, species <span class="op">==</span><span class="st"> "Homo sapiens"</span>)
<span class="co">#----------------------------------------------------------------------------------------------------</span>
<span class="co"># add refseq promoters, available from RefSeq for each transcript which has been identified</span>
<span class="co">#----------------------------------------------------------------------------------------------------</span>
ah.human.refseq &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/MotifDb/man/query.html">query</a></span>(ah.human, <span class="st">"RefSeq"</span>, <span class="st">"hg19"</span>, <span class="st">"RefSeq Genes"</span>)

<span class="co"># download the first set</span>
human.refseq &lt;-<span class="st"> </span>ah.human.refseq[[<span class="dv">1</span>]]
gr.promoters &lt;-<span class="st"> </span><span class="kw">promoters</span>(human.refseq, <span class="dt">upstream=</span><span class="dv">2000</span>, <span class="dt">downstream=</span><span class="dv">200</span>)
  <span class="co"># get rid of score, itemRgb, thick, blocks columns in the mcols, keeping just the transcript name.</span>
  <span class="co"># these attributes are meaningful for transcript tracks since those include the represenation</span>
  <span class="co"># of UTRs, introns and exons.   but a promoter is a stretch of DNA for which those distinctions</span>
  <span class="co"># do not apply</span>
<span class="kw">mcols</span>(gr.promoters) &lt;-<span class="st"> </span><span class="kw">mcols</span>(gr.promoters)[,<span class="dv">1</span>]
<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw">mcols</span>(gr.promoters)) &lt;-<span class="st"> "name"</span>
ov &lt;-<span class="st"> </span><span class="kw">findOverlaps</span>(gr.promoters, mef2c.region)
gr.mef2c.promoters &lt;-<span class="st"> </span>gr.promoters[<span class="kw">queryHits</span>(ov)]
track.promoters &lt;-<span class="st"> </span><span class="kw"><a href="../reference/UCSCBedAnnotationTrack-class.html">UCSCBedAnnotationTrack</a></span>(<span class="st">"promoters"</span>, gr.mef2c.promoters, <span class="dt">color=</span><span class="st">"darkGreen"</span>)
<span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track.promoters)</code></pre></div>
</div>
<div id="overlap-promoters-and-variants-create-new-track" class="section level1">
<h1 class="hasAnchor">
<a href="#overlap-promoters-and-variants-create-new-track" class="anchor"></a>Overlap promoters and variants, create new track</h1>
<p>Non-coding SNPs which fall in promoter regions are plausible candidates for further examination. They may disrupt or create a transcription factor binding site, and thus possibly contribute to the disease under study.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gr.variants &lt;-<span class="st"> </span><span class="kw">GRanges</span>(tbl.mef2cGWAS.variants.bed)
ov &lt;-<span class="st"> </span><span class="kw">findOverlaps</span>(gr.variants, gr.promoters)
gr.variantsInPromoters &lt;-<span class="st"> </span>gr.variants[<span class="kw">queryHits</span>(ov)]
track.variantsInPromoters &lt;-<span class="kw"><a href="../reference/GRangesAnnotationTrack-class.html">GRangesAnnotationTrack</a></span>(<span class="st">"snpInPromoters"</span>, gr.variantsInPromoters,
                                                   <span class="dt">color=</span><span class="st">"red"</span>, <span class="dt">displayMode=</span><span class="st">"EXPANDED"</span>)
<span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track.variantsInPromoters)</code></pre></div>
</div>
<div id="add-methylation-tracks" class="section level1">
<h1 class="hasAnchor">
<a href="#add-methylation-tracks" class="anchor"></a>Add Methylation Tracks</h1>
<p>H3K4me3 histone methylation marks are often associated with transcription initiation in promoter regions. We will add three tracks from the AnnotationHub for H3K4me3 methylation in <a href="https://www.genome.gov/26524238/encode-project-common-cell-types/">Gm12878</a> (lymphoblastoid) cells.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">histone.tracks &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/MotifDb/man/query.html">query</a></span>(ah.human, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"H3K4me3"</span>, <span class="st">"Gm12878"</span>, <span class="st">"Peak"</span>, <span class="st">"narrow"</span>))  <span class="co"># 3 tracks</span>
descriptions &lt;-<span class="st"> </span>histone.tracks<span class="op">$</span>description
titles &lt;-<span class="st"> </span>histone.tracks<span class="op">$</span>title
colors &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/grDevices/palettes.html">terrain.colors</a></span>(<span class="dv">6</span>), <span class="dv">4</span>)
color.index &lt;-<span class="st"> </span><span class="dv">0</span>

<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(histone.tracks))){
   name &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(histone.tracks)[i]
   color.index &lt;-<span class="st"> </span>color.index <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
   gr &lt;-<span class="st"> </span>histone.tracks[[name]]
   ov &lt;-<span class="st"> </span><span class="kw">findOverlaps</span>(gr, mef2c.region)
   mef2c.histones &lt;-<span class="st"> </span>gr[<span class="kw">queryHits</span>(ov)]
   track.histones &lt;-<span class="st"> </span><span class="kw"><a href="../reference/GRangesQuantitativeTrack-class.html">GRangesQuantitativeTrack</a></span>(titles[i], mef2c.histones[, <span class="st">"pValue"</span>],
                                              <span class="dt">color=</span>colors[color.index], <span class="dt">trackHeight=</span><span class="dv">50</span>,
                                              <span class="dt">autoscale=</span><span class="ot">TRUE</span>)
   <span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track.histones)
   <span class="kw"><a href="https://rdrr.io/r/base/Sys.sleep.html">Sys.sleep</a></span>(<span class="dv">5</span>)
   } <span class="co"># for track</span></code></pre></div>
</div>
<div id="vcf-from-adni-the-alzheimers-disease-neuroimaging-initiative" class="section level1">
<h1 class="hasAnchor">
<a href="#vcf-from-adni-the-alzheimers-disease-neuroimaging-initiative" class="anchor"></a>VCF from ADNI, the Alzheimer’s Disease Neuroimaging Initiative</h1>
<p>We include a very small slice of the ADNI whole genome sequencing <a href="http://adni.loni.usc.edu/data-samples/genetic-data/wgs/">data</a>. 349 samples (filtered from a larger set of 128 AD, 415 Mild Cognitive Impairment, 267 controls, 8 uncertains) are included, all in a ~4kb region including methylation marks, and IGAP GWAS SNPs in promoter regions.</p>
<p>In order to illustrate some of genomic data exploration capabilities of igvR we are therefore mixing apples and oranges. We have GWAS variants from IGAP (a very large meta-analysis), promoters from RefSeq (conservatively defined), methylation marks from a lymphoblastoid cell line, and WGS variants from ADNI. Note also that the ADNI variants in the VCF track are not stratified by disease phenotype. Therefore, examining rsid rs80043958 at chr5:88,179,576 with an IGAP GWAS pvalue of 0.002, we see the ADNI heterozygosity is 0.269 - this and other information about that variant is displayed in a popup by clicking on the pinkish “anchor block” - keep in mind that that number would be more informative if broken out by phenotype.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vcfFilename &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="dt">package=</span><span class="st">"igvR"</span>, <span class="st">"extdata"</span>, <span class="st">"mef2c-4kb.vcf"</span>)
vcf &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html">readVcf</a></span>(vcfFilename, <span class="st">"hg19"</span>)
track.vcf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/VariantTrack-class.html">VariantTrack</a></span>(<span class="st">"AMPAD VCF"</span>, vcf, <span class="dt">trackHeight=</span><span class="dv">1000</span>)
<span class="kw"><a href="https://rdrr.io/r/base/Sys.sleep.html">Sys.sleep</a></span>(<span class="dv">5</span>)
<span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track.vcf)
<span class="kw"><a href="https://rdrr.io/r/base/Sys.sleep.html">Sys.sleep</a></span>(<span class="dv">10</span>)  <span class="co"># allow plenty of time for this to complete, and the track to appear, before ending this chunk</span></code></pre></div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<ol style="list-style-type: decimal">
<li>Lambert, Jean-Charles, et al. “Meta-analysis of 74,046 individuals identifies 11 new susceptibility loci for Alzheimer’s disease.” <a href="https://www.ncbi.nlm.nih.gov/pubmed/24162737">Nature genetics 45.12 (2013): 1452</a>.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#create-a-granges-object-specifying-our-region-of-interest-285-kb-centered-on-the-mef2c-gene">Create a GRanges object specifying our region of interest, 285 kb centered on the MEF2C gene</a></li>
      <li><a href="#use-the-annotationhub-to-create-a-promoter-track">Use the AnnotationHub to create a promoter track</a></li>
      <li><a href="#overlap-promoters-and-variants-create-new-track">Overlap promoters and variants, create new track</a></li>
      <li><a href="#add-methylation-tracks">Add Methylation Tracks</a></li>
      <li><a href="#vcf-from-adni-the-alzheimers-disease-neuroimaging-initiative">VCF from ADNI, the Alzheimer’s Disease Neuroimaging Initiative</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
