<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Explore ChIP-seq alignments from a bam file, MACS2 narrowPeaks, and  motif matching, near GATA2 • igvR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Explore ChIP-seq alignments from a bam file, MACS2 narrowPeaks, and  motif matching, near GATA2">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">igvR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.37</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/alzheimersVariantsNearMEF2C.html">Explore VCF variants, GWAS snps, promoters and histone marks around the  MEF2C gene in Alzheimers Disease</a>
    </li>
    <li>
      <a href="../articles/basicIntro.html">Introduction: a simple demo</a>
    </li>
    <li>
      <a href="../articles/ctcfChipSeq.html">Explore ChIP-seq alignments from a bam file, MACS2 narrowPeaks, and  motif matching, near GATA2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Explore ChIP-seq alignments from a bam file, MACS2 narrowPeaks, and motif matching, near GATA2</h1>
                        <h4 class="author">Paul Shannon</h4>
            
            <h4 class="date">2019-10-17</h4>
      
      
      <div class="hidden name"><code>ctcfChipSeq.Rmd</code></div>

    </div>

    
    
<style>
.main-container { width: 1200px; max-width:2800px;}
</style>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>The igvR package provides easy programmic access in R to the web-based javascript library <a href="https://github.com/igvteam/igv.js">igv.js</a> in order to create and display genome tracks in its richly interactive web browser visual interface. I am grateful to Jim Robinson, Douglass Turner and colleagues for their fine work.</p>
<p>In this vignette we present a somewhat contrived ChIP-seq study - contrived in that it is not organized around an actual research problem. Instead, this vignette demonstrates methods you would likely use to do visual QC and exploaratory data analysis a ChIP-seq experement.</p>
<p>We begin with an ENCODE study of CTCF binding on chromosome 3, hg19 reference genome, in the vicinity of the GATA2 gene. We will</p>
<ul>
<li>display a 252 kb region of that genome</li>
<li>load and display a GenomicAlignments object from an indexed bam file covering that region</li>
<li>load and display a narrow peaks bedGraph track, precalculated by MACS2</li>
<li>Use the Biostrings::matchPWM method to identify regions in the displayed sequence which, with at least 80% fidelity, match the CTCF motif, Jaspar 2018’s MA0138.1, from MotifDb</li>
<li>Display that motif logo in a browser popup.</li>
</ul>
<p>This screenshot illustrates the final state of igv in your browser at the conclusion of the code found below.</p>
<p><img src="ctcf-chip-seq-gata2-252kb-igvR.png" width="824"></p>
</div>
<div id="load-the-libraries-we-need" class="section level1">
<h1 class="hasAnchor">
<a href="#load-the-libraries-we-need" class="anchor"></a>Load the libraries we need</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(igvR)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(MotifDb)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BSgenome.Hsapiens.UCSC.hg19)</code></pre></div>
<p>Create the igvR instance, with all default parameters (portRange, quiet, title). Javascript and HTML is loaded into your browser, igv.js is initialized, a websocket connection between your R process and that web page is constructed, over which subsequent commands and data will travel.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">igv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/igvR-class.html">igvR</a></span>()
<span class="kw">setBrowserWindowTitle</span>(igv, <span class="st">"CTCF ChIP-seq"</span>)
<span class="kw"><a href="../reference/setGenome.html">setGenome</a></span>(igv, <span class="st">"hg19"</span>)</code></pre></div>
<p>Display 1.4MB on chr3 more or less centered on the GATA2 genes</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/showGenomicRegion.html">showGenomicRegion</a></span>(igv, <span class="st">"chr3:128,079,020-128,331,275"</span>)</code></pre></div>
</div>
<div id="create-a-granges-object-specifying-our-region-of-interest-obtaining-a-list-chrom-start-end-string-from-igv" class="section level1">
<h1 class="hasAnchor">
<a href="#create-a-granges-object-specifying-our-region-of-interest-obtaining-a-list-chrom-start-end-string-from-igv" class="anchor"></a>Create a GRanges object specifying our region of interest, obtaining a list (chrom, start, end, string) from igv</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">loc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGenomicRegion.html">getGenomicRegion</a></span>(igv)
which &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/with.html">with</a></span>(loc, <span class="kw">GRanges</span>(<span class="dt">seqnames=</span>chrom, <span class="dt">ranges =</span> <span class="kw">IRanges</span>(start, end)))
param &lt;-<span class="st"> </span><span class="kw">ScanBamParam</span>(<span class="dt">which=</span>which, <span class="dt">what =</span> <span class="kw">scanBamWhat</span>())
bamFile &lt;-<span class="st"> "~/github/ChIPseqMotifMatch/bulk/GSM749704/GSM749704_hg19_wgEncodeUwTfbsGm12878CtcfStdAlnRep1.bam"</span>
<span class="kw"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span>(bamFile)
x &lt;-<span class="st"> </span><span class="kw">readGAlignments</span>(bamFile, <span class="dt">use.names=</span><span class="ot">TRUE</span>, <span class="dt">param=</span>param)
track &lt;-<span class="st"> </span><span class="kw"><a href="../reference/GenomicAlignmentTrack-class.html">GenomicAlignmentTrack</a></span>(<span class="st">"ctcf bam"</span>, x, <span class="dt">visibilityWindow=</span><span class="dv">10000000</span>, <span class="dt">trackHeight=</span><span class="dv">200</span>)  <span class="co"># 30000 default</span>
<span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track)</code></pre></div>
</div>
<div id="load-narrow-peaks-called-by-macs2-from-the-bam-file" class="section level1">
<h1 class="hasAnchor">
<a href="#load-narrow-peaks-called-by-macs2-from-the-bam-file" class="anchor"></a>Load narrow peaks called by MACS2 from the bam file</h1>
<p>We ran MACS2 using docker and make, with this makefile:</p>
<pre><code>DATADIR=/Users/paul/github/igvR/vignettes/macs2
BAM=GSM749704_hg19_wgEncodeUwTfbsGm12878CtcfStdAlnRep1.bam
NAME=GSM749704_hg19_chr19
run:
    docker run -v $(DATADIR):/data/ fooliu/macs2 callpeak -t /data/$(BAM) -n $(NAME) --outdir data/</code></pre>
<p>This produces several files; we are interested only in the narrow peaks:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filename &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="dt">package=</span><span class="st">"igvR"</span>, <span class="st">"extdata"</span>, <span class="st">"GSM749704_hg19_chr19_peaks.narrowPeak"</span>)
tbl.pk &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(filename, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span><span class="ot">FALSE</span>, <span class="dt">as.is=</span><span class="ot">TRUE</span>)[, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">5</span>)]
<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(tbl.pk)
<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(tbl.pk) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chrom"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"score"</span>)
loc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGenomicRegion.html">getGenomicRegion</a></span>(igv)
<span class="cf">while</span>(<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">is.list</a></span>(loc)){
   <span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(loc)
   loc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGenomicRegion.html">getGenomicRegion</a></span>(igv)
   }
<span class="kw">printf</span>(<span class="st">"--- good loc obtained? "</span>)
<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(loc)

tbl.pk &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(tbl.pk, chrom<span class="op">==</span>loc<span class="op">$</span>chrom <span class="op">&amp;</span><span class="st"> </span>start <span class="op">&gt;=</span><span class="st"> </span>loc<span class="op">$</span>start <span class="op">&amp;</span><span class="st"> </span>end <span class="op">&lt;=</span><span class="st"> </span>loc<span class="op">$</span>end)
<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(tbl.pk)
track &lt;-<span class="st"> </span><span class="kw"><a href="../reference/DataFrameQuantitativeTrack-class.html">DataFrameQuantitativeTrack</a></span>(<span class="st">"macs2 peaks"</span>, tbl.pk, <span class="dt">color=</span><span class="st">"red"</span>, <span class="dt">autoscale=</span><span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track)</code></pre></div>
</div>
<div id="find-and-display-motifs-for-ctcf" class="section level1">
<h1 class="hasAnchor">
<a href="#find-and-display-motifs-for-ctcf" class="anchor"></a>Find and display motifs for CTCF</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
dna &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/with.html">with</a></span>(loc, <span class="kw">getSeq</span>(BSgenome.Hsapiens.UCSC.hg19, chrom, start, end))


pfm.ctcf &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/MotifDb/man/query.html">query</a></span>(MotifDb, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CTCF"</span>, <span class="st">"sapiens"</span>, <span class="st">"jaspar2018"</span>), <span class="dt">notStrings=</span><span class="st">"ctcfl"</span>)
motif.name &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(pfm.ctcf)[<span class="dv">1</span>]
pfm &lt;-<span class="st"> </span>pfm.ctcf[[<span class="dv">1</span>]]

hits.forward &lt;-<span class="st"> </span><span class="kw">matchPWM</span>(pfm, <span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(dna), <span class="dt">with.score=</span><span class="ot">TRUE</span>, <span class="dt">min.score=</span><span class="st">"80%"</span>)
hits.reverse &lt;-<span class="st"> </span><span class="kw">matchPWM</span>(<span class="kw">reverseComplement</span>(pfm), <span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(dna), <span class="dt">with.score=</span><span class="ot">TRUE</span>, <span class="dt">min.score=</span><span class="st">"80%"</span>)

tbl.forward &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="kw">ranges</span>(hits.forward))
tbl.reverse &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="kw">ranges</span>(hits.reverse))
tbl.forward<span class="op">$</span>score &lt;-<span class="st"> </span><span class="kw">mcols</span>(hits.forward)<span class="op">$</span>score
tbl.reverse<span class="op">$</span>score &lt;-<span class="st"> </span><span class="kw">mcols</span>(hits.reverse)<span class="op">$</span>score

tbl.matches &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(tbl.forward, tbl.reverse)


tbl.matches<span class="op">$</span>chrom &lt;-<span class="st"> </span>loc<span class="op">$</span>chrom
tbl.matches<span class="op">$</span>start &lt;-<span class="st"> </span>tbl.matches<span class="op">$</span>start <span class="op">+</span><span class="st"> </span>loc<span class="op">$</span>start
tbl.matches<span class="op">$</span>end &lt;-<span class="st"> </span>tbl.matches<span class="op">$</span>end <span class="op">+</span><span class="st"> </span>loc<span class="op">$</span>start
<span class="kw"><a href="https://rdrr.io/r/base/with.html">with</a></span>(tbl.matches, end<span class="op">-</span>start)

<span class="kw"><a href="../reference/enableMotifLogoPopups.html">enableMotifLogoPopups</a></span>(igv)

tbl.matches<span class="op">$</span>name &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"MotifDb::"</span>, motif.name)
tbl.matches &lt;-<span class="st"> </span>tbl.matches[, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chrom"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"name"</span>, <span class="st">"score"</span>)]
<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(tbl.matches)

<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(tbl.matches)
track &lt;-<span class="st"> </span><span class="kw"><a href="../reference/DataFrameAnnotationTrack-class.html">DataFrameAnnotationTrack</a></span>(<span class="st">"CTCF-MA0139.1"</span>, tbl.matches, <span class="dt">color=</span><span class="st">"random"</span>)
<span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track)

track &lt;-<span class="st"> </span><span class="kw"><a href="../reference/DataFrameQuantitativeTrack-class.html">DataFrameQuantitativeTrack</a></span>(<span class="st">"MA0139.1 scores"</span>, tbl.matches[, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">5</span>)], <span class="dt">color=</span><span class="st">"random"</span>, <span class="dt">autoscale=</span><span class="ot">FALSE</span>, <span class="dt">min=</span><span class="dv">8</span>, <span class="dt">max=</span><span class="dv">12</span>)
<span class="kw"><a href="../reference/displayTrack.html">displayTrack</a></span>(igv, track)</code></pre></div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#load-the-libraries-we-need">Load the libraries we need</a></li>
      <li><a href="#create-a-granges-object-specifying-our-region-of-interest-obtaining-a-list-chrom-start-end-string-from-igv">Create a GRanges object specifying our region of interest, obtaining a list (chrom, start, end, string) from igv</a></li>
      <li><a href="#load-narrow-peaks-called-by-macs2-from-the-bam-file">Load narrow peaks called by MACS2 from the bam file</a></li>
      <li><a href="#find-and-display-motifs-for-ctcf">Find and display motifs for CTCF</a></li>
      <li><a href="#session-info">Session Info</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
