<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="igvR">
<title>Explore CTCF ChIP-seq alignments, MACS2 narrowPeaks, Motif Matching and H3K4me3 methylation • igvR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Explore CTCF ChIP-seq alignments, MACS2 narrowPeaks, Motif Matching and H3K4me3 methylation">
<meta property="og:description" content="igvR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">igvR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.17.21</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/v00.basicIntro.html">Introduction: a simple demo</a>
    <a class="dropdown-item" href="../articles/v01.stockGenome.html">Use a Stock Genome</a>
    <a class="dropdown-item" href="../articles/v02.customGenome.html">Use a Custom Genome</a>
    <a class="dropdown-item" href="../articles/v03.ctcfChIP.html">Explore CTCF ChIP-seq alignments, MACS2 narrowPeaks, Motif Matching and H3K4me3 methylation</a>
    <a class="dropdown-item" href="../articles/v04.pairedEnd.html">Paired-end Interaction Tracks</a>
    <a class="dropdown-item" href="../articles/v05.ucscTableBrowser.html">Obtain and Display H3K4Me3 K562 track from UCSC table browser</a>
    <a class="dropdown-item" href="../articles/v06.annotationHub.html">Obtain and Display H3K27ac K562 track from the AnnotationHub</a>
    <a class="dropdown-item" href="../articles/v07.gwas.html">GWAS Tracks</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="v03.ctcfChIP_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Explore CTCF ChIP-seq alignments, MACS2 narrowPeaks, Motif Matching and H3K4me3 methylation</h1>
                        <h4 data-toc-skip class="author">Paul Shannon</h4>
            
            <h4 data-toc-skip class="date">2022-08-05</h4>
      
      
      <div class="d-none name"><code>v03.ctcfChIP.Rmd</code></div>
    </div>

    
    
<style>
.main-container { width: 1200px; max-width:2800px;}
</style>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The igvR package provides easy programmatic access in R to the web-based javascript library <a href="https://github.com/igvteam/igv.js" class="external-link">igv.js</a>. R’s access to data, and operations upon data, are complemented by igv’s richly interactive web browser interface to display and explore genomic and epigentic data.</p>
<p>In this vignette we present a somewhat contrived ChIP-seq study - contrived in that it is not organized around an actual research problem. Instead, this vignette demonstrates methods you would likely use to do visual QC and exploratory data analysis of ChIP-seq data, for CTCF, a <a href="https://www.nature.com/articles/emm201533#:~:text=CCCTC%2Dbinding%20factor%20(CTCF),communication%20between%20enhancers%20and%20promoters" class="external-link">protein which plays multiple roles in gene regulation</a> in the vicinity of a gene of interest, GATA2.</p>
<p>We begin with ENCODE data of CTCF binding on chromosome 3, on the hg19 reference genome, in the vicinity of the GATA2 gene. We will</p>
<ul>
<li>Display a 252 kb region of that genome</li>
<li>Read the indexed bam file into a GenomicAlignments object covering that region, and display it</li>
<li>Load and display a narrow peaks bedGraph track, precalculated by MACS2 processing of the bam file</li>
<li>Use the Biostrings::matchPWM method to identify regions in the displayed sequence which, with at least 80% fidelity, match the CTCF motif, Jaspar 2018’s MA0138.1, from MotifDb</li>
<li>Display those scored motif matches</li>
<li>Load and display H3K4me3 histone marks from a bigwig file</li>
<li>Zoom in on a likely functional CTCF binding site downstream of GATA2, as suggested by these combined data</li>
</ul>
</div>
<div class="section level2">
<h2 id="initialize-igvr">Initialize igvR<a class="anchor" aria-label="anchor" href="#initialize-igvr"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://paul-shannon.github.io/igvR/" class="external-link">igvR</a></span><span class="op">)</span></span>
<span><span class="va">igv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/igvR-class.html">igvR</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://paul-shannon.github.io/BrowserViz/reference/setBrowserWindowTitle.html" class="external-link">setBrowserWindowTitle</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"CTCF ChIP-seq"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/setGenome.html">setGenome</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"hg19"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/showGenomicRegion.html">showGenomicRegion</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"chr3:128,079,020-128,331,275"</span><span class="op">)</span></span>
<span>   <span class="co"># or</span></span>
<span><span class="fu"><a href="../reference/showGenomicRegion.html">showGenomicRegion</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"GATA2"</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span> <span class="fu"><a href="../reference/zoomOut.html">zoomOut</a></span><span class="op">(</span><span class="va">igv</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/ctcfBam-01.png" width="474"></p>
</div>
<div class="section level2">
<h2 id="load-alignments-for-the-currently-displayed-genomic-region">Load alignments for the currently displayed genomic region<a class="anchor" aria-label="anchor" href="#load-alignments-for-the-currently-displayed-genomic-region"></a>
</h2>
<p>A modest but often useful feature of igvR is R access to the currently displayed genomic region. In the previous step we showed to ways to set the region programatically from R. This may not get you exactly the region you are interested in, so interactive adjustment in the browser can fine tune your region of interest (“roi”). Once that is set, retrieve the roi, create a GRanges object from it to use to read only the relevant slice of the bam file.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">roi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGenomicRegion.html">getGenomicRegion</a></span><span class="op">(</span><span class="va">igv</span><span class="op">)</span></span>
<span><span class="va">gr.roi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">roi</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames<span class="op">=</span><span class="va">chrom</span>, ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span><span class="va">start</span>, <span class="va">end</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu">ScanBamParam</span><span class="op">(</span>which<span class="op">=</span><span class="va">gr.roi</span>, what <span class="op">=</span> <span class="fu">scanBamWhat</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bamFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"igvR"</span>, <span class="st">"extdata"</span>, <span class="st">"ctcf-gata2"</span>, <span class="st">"gata2-region-hg19.bam"</span><span class="op">)</span></span>
<span><span class="va">alignments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html" class="external-link">readGAlignments</a></span><span class="op">(</span><span class="va">bamFile</span>, use.names<span class="op">=</span><span class="cn">TRUE</span>, param<span class="op">=</span><span class="va">param</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="display-the-alignment">Display the alignment<a class="anchor" aria-label="anchor" href="#display-the-alignment"></a>
</h2>
<p>igvR provides a number of tracks, each with a constructor, each which then is displayed in igv with the same general purpose function. These are some of the currently supported tracks:</p>
<ul>
<li>BedpeInteractionsTrack</li>
<li>DataFrameAnnotationTrack</li>
<li>DataFrameQuantitativeTrack</li>
<li>GFF3Track</li>
<li>GRangesAnnotationTrack</li>
<li>GRangesQuantitativeTrack</li>
<li>GWASTrack</li>
<li>GWASUrlTrack</li>
<li>GenomicAlignmentTrack</li>
<li>RemoteAlignmentTrack</li>
<li>UCSCBedAnnotationTrack</li>
<li>UCSCBedGraphQuantitativeTrack</li>
<li>VariantTrack</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GenomicAlignmentTrack-class.html">GenomicAlignmentTrack</a></span><span class="op">(</span>trackName<span class="op">=</span><span class="st">"ctcf bam"</span>, <span class="va">alignments</span>, visibilityWindow<span class="op">=</span><span class="fl">10000000</span>, trackHeight<span class="op">=</span><span class="fl">200</span><span class="op">)</span> </span>
<span><span class="fu"><a href="../reference/displayTrack.html">displayTrack</a></span><span class="op">(</span><span class="va">igv</span>, <span class="va">track</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/ctcfBam-02.png" width="678"></p>
</div>
<div class="section level2">
<h2 id="narrow-peaks-from-macs">Narrow Peaks from MACS<a class="anchor" aria-label="anchor" href="#narrow-peaks-from-macs"></a>
</h2>
<p><a href="https://github.com/macs3-project/MACS" class="external-link">MACS</a> is a popular peak-caller, summarizing piled-up DNA reads into narrow peaks. This operation sacrifices detail, and is subject to run-time parameterization choices, but is nonetheless quite useful. By examinging these narrow peaks and the raw reads together in igvR, the benefits of detail and summary are both achieved.</p>
<p>The <em>DataFrameQuantitativeTrack</em> requires a data.frame with the columns and classes shown here:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">lapply</span>(tbl.pk, class)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="kw">head</span>(tbl.pk)</span>
<span id="cb4-3"><a href="#cb4-3"></a></span>
<span id="cb4-4"><a href="#cb4-4"></a>constructed from a data.frame.  </span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">narrowPeaksFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"igvR"</span>, <span class="st">"extdata"</span>, <span class="st">"ctcf-gata2"</span>,</span>
<span>                               <span class="st">"gata2-region-macs2-narrowPeaks.RData"</span><span class="op">)</span></span>
<span><span class="va">tbl.pk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html" class="external-link">get</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">narrowPeaksFile</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">tbl.pk</span><span class="op">)</span> <span class="co"># 109 4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tbl.pk</span><span class="op">)</span></span>
<span>  <span class="co">#      chrom     start       end score</span></span>
<span>  <span class="co"># 6381 chr10 127910682 127910864    27</span></span>
<span>  <span class="co"># 6382 chr10 128075644 128075811    89</span></span>
<span>  <span class="co"># 6383 chr10 128259852 128259984    27</span></span>
<span>  <span class="co"># 6384 chr10 128286655 128286920    78</span></span>
<span>  <span class="co"># 6385 chr10 128437706 128437938    89</span></span>
<span>  <span class="co"># 8827 chr11 127965327 127965489    70</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">tbl.pk</span>, <span class="va">class</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#        chrom       start         end       score </span></span>
<span><span class="co">#  "character"   "integer"   "integer"   "integer" </span></span>
<span></span>
<span><span class="va">track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DataFrameQuantitativeTrack-class.html">DataFrameQuantitativeTrack</a></span><span class="op">(</span><span class="st">"macs2 peaks"</span>, <span class="va">tbl.pk</span>, color<span class="op">=</span><span class="st">"red"</span>, autoscale<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/displayTrack.html">displayTrack</a></span><span class="op">(</span><span class="va">igv</span>, <span class="va">track</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/ctcfBam-03.png" width="679"></p>
</div>
<div class="section level2">
<h2 id="motif-matching">Motif Matching<a class="anchor" aria-label="anchor" href="#motif-matching"></a>
</h2>
<p>Transcription factor binding sites are often characterized by sequence match to the TF’s curated motif. We will use this motif from <a href="https://jaspar.genereg.net/matrix/MA0139.1/" class="external-link">JASPAR 2022</a> and identify matches in the hg19 genome in our region of interest.</p>
<p><img src="images/ctcfMotif.png" width="500px"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>   <span class="co"># get the DNA sequence in the current region</span></span>
<span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">roi</span>, <span class="fu">getSeq</span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span>, <span class="va">chrom</span>, <span class="va">start</span>, <span class="va">end</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="co"># get the first of three motifs for CTCF.  (a more thorough study would use all three)</span></span>
<span><span class="va">pfm.ctcf</span> <span class="op">&lt;-</span> <span class="fu">MotifDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MotifDb/man/query.html" class="external-link">query</a></span><span class="op">(</span><span class="va">MotifDb</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTCF"</span>, <span class="st">"sapiens"</span>, <span class="st">"jaspar2022"</span><span class="op">)</span>, notStrings<span class="op">=</span><span class="st">"ctcfl"</span><span class="op">)</span></span>
<span><span class="va">motif.name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pfm.ctcf</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">pfm</span> <span class="op">&lt;-</span> <span class="va">pfm.ctcf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span>   <span class="co"># Find matches &gt;= 80% of this motif in the sequence.  create a suitable </span></span>
<span>   <span class="co"># data.frame for another DataFrameQuantitativeTrack</span></span>
<span></span>
<span><span class="va">hits.forward</span> <span class="op">&lt;-</span> <span class="fu">matchPWM</span><span class="op">(</span><span class="va">pfm</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span>, with.score<span class="op">=</span><span class="cn">TRUE</span>, min.score<span class="op">=</span><span class="st">"80%"</span><span class="op">)</span></span>
<span><span class="va">hits.reverse</span> <span class="op">&lt;-</span> <span class="fu">matchPWM</span><span class="op">(</span><span class="fu">reverseComplement</span><span class="op">(</span><span class="va">pfm</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span>, with.score<span class="op">=</span><span class="cn">TRUE</span>, min.score<span class="op">=</span><span class="st">"80%"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tbl.forward</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">ranges</span><span class="op">(</span><span class="va">hits.forward</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tbl.reverse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">ranges</span><span class="op">(</span><span class="va">hits.reverse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tbl.forward</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">hits.forward</span><span class="op">)</span><span class="op">$</span><span class="va">score</span></span>
<span><span class="va">tbl.reverse</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">hits.reverse</span><span class="op">)</span><span class="op">$</span><span class="va">score</span></span>
<span></span>
<span><span class="va">tbl.matches</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">tbl.forward</span>, <span class="va">tbl.reverse</span><span class="op">)</span></span>
<span><span class="va">tbl.matches</span><span class="op">$</span><span class="va">chrom</span> <span class="op">&lt;-</span> <span class="va">roi</span><span class="op">$</span><span class="va">chrom</span></span>
<span><span class="va">tbl.matches</span><span class="op">$</span><span class="va">start</span> <span class="op">&lt;-</span> <span class="va">tbl.matches</span><span class="op">$</span><span class="va">start</span> <span class="op">+</span> <span class="va">roi</span><span class="op">$</span><span class="va">start</span></span>
<span></span>
<span><span class="va">tbl.matches</span><span class="op">$</span><span class="va">end</span> <span class="op">&lt;-</span> <span class="va">tbl.matches</span><span class="op">$</span><span class="va">end</span> <span class="op">+</span> <span class="va">roi</span><span class="op">$</span><span class="va">start</span></span>
<span></span>
<span><span class="va">tbl.matches</span><span class="op">$</span><span class="va">name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"MotifDb::"</span>, <span class="va">motif.name</span><span class="op">)</span></span>
<span><span class="va">tbl.matches</span> <span class="op">&lt;-</span> <span class="va">tbl.matches</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrom"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"name"</span>, <span class="st">"score"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">tbl.matches</span><span class="op">)</span> <span class="co"># 25 5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tbl.matches</span><span class="op">)</span></span>
<span>    <span class="co">#   chrom     start       end                                       name    score</span></span>
<span>    <span class="co"># 1  chr3 128110910 128110928 MotifDb::Hsapiens-jaspar2018-CTCF-MA0139.1 10.70369</span></span>
<span>    <span class="co"># 2  chr3 128114573 128114591 MotifDb::Hsapiens-jaspar2018-CTCF-MA0139.1 10.36891</span></span>
<span>    <span class="co"># 3  chr3 128127658 128127676 MotifDb::Hsapiens-jaspar2018-CTCF-MA0139.1 10.44666</span></span>
<span>    <span class="co"># 4  chr3 128138376 128138394 MotifDb::Hsapiens-jaspar2018-CTCF-MA0139.1 10.54861</span></span>
<span>    <span class="co"># 5  chr3 128139280 128139298 MotifDb::Hsapiens-jaspar2018-CTCF-MA0139.1 10.51689</span></span>
<span>    <span class="co"># 6  chr3 128173128 128173146 MotifDb::Hsapiens-jaspar2018-CTCF-MA0139.1 10.37987</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="display-the-matches">Display the Matches<a class="anchor" aria-label="anchor" href="#display-the-matches"></a>
</h2>
<p>This will show that there are many more potential binding sites, based on sequence alone, that is indicated by ChIP.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DataFrameQuantitativeTrack-class.html">DataFrameQuantitativeTrack</a></span><span class="op">(</span><span class="st">"MA0139.1 scores"</span>, <span class="va">tbl.matches</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                                    color<span class="op">=</span><span class="st">"random"</span>, autoscale<span class="op">=</span><span class="cn">FALSE</span>, min<span class="op">=</span><span class="fl">8</span>, max<span class="op">=</span><span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/displayTrack.html">displayTrack</a></span><span class="op">(</span><span class="va">igv</span>, <span class="va">track</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/ctcfBam-04.png" width="680"></p>
</div>
<div class="section level2">
<h2 id="h3k4me3-histone-marks">H3K4Me3 Histone Marks<a class="anchor" aria-label="anchor" href="#h3k4me3-histone-marks"></a>
</h2>
<p>Methylation at the 4th lysine of histone H3 can correlate with transcrition factor binding nearby. Here we read and load an H3k4Me3 track from the same <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM749704" class="external-link">GM12878</a> lymphoblastoid cells from which the CTCF ChIP-seq was obtained.</p>
<p>The Bioconductor AnnotationHub is one possible source for this H3K4Me3 data. We obtained this track, <em>AH40294 | E116-H3K4me3.imputed.pval.signal.bigwig</em>, and then for convenience sliced out a region of intererest and saved as a bigwig file in extdata/ctcf-gata2 directory of the igvR package.</p>
<p>We use here the <em>GRangesQuantitativeTrack</em>, which is very like the <em>DataFrameQuantitativeTrack</em> used above, but specialized for GRanges rather than bed-like data.frames.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">roi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGenomicRegion.html">getGenomicRegion</a></span><span class="op">(</span><span class="va">igv</span><span class="op">)</span></span>
<span><span class="va">bigwig.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"igvR"</span>, <span class="st">"extdata"</span>, <span class="st">"ctcf-gata2"</span>, <span class="st">"gata2-region-h3k4me3.bw"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bw.slice</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">bigwig.file</span>, which<span class="op">=</span><span class="va">gr.roi</span><span class="op">)</span></span>
<span><span class="va">track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRangesQuantitativeTrack-class.html">GRangesQuantitativeTrack</a></span><span class="op">(</span><span class="st">"h3k4me3"</span>, <span class="va">bw.slice</span>, autoscale<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/displayTrack.html">displayTrack</a></span><span class="op">(</span><span class="va">igv</span>, <span class="va">track</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/ctcfBam-05.png" width="681"></p>
</div>
<div class="section level2">
<h2 id="zoom-in-to-one-interesting-and-possibly-functional-ctcf-binding-site">Zoom in to one interesting, and possibly functional CTCF binding site<a class="anchor" aria-label="anchor" href="#zoom-in-to-one-interesting-and-possibly-functional-ctcf-binding-site"></a>
</h2>
<p>This is most easily, and most naturally accomplished with your mouse in the web browser, by dragging a region in the spece between the cytoband and the sequence ruler, producing the zoomed in view which can also be obtained programmatically:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/showGenomicRegion.html">showGenomicRegion</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"chr3:128,202,505-128,222,868"</span><span class="op">)</span></span></code></pre></div>
<p>You can see that the ChIP-seq evidence supports CTCF binding, with additional support from motif match and H3K4Me3. There is no ChIP-seq support for the upstream site, despite motif match and H3K4Me3 binding.</p>
<p><img src="images/ctcfBam-07.png" width="678"></p>
<!--
  # Session Info
  
  
  ```r
  sessionInfo()
  #> R version 4.2.0 (2022-04-22)
  #> Platform: x86_64-apple-darwin17.0 (64-bit)
  #> Running under: macOS Big Sur/Monterey 10.16
  #> 
  #> Matrix products: default
  #> BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
  #> LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib
  #> 
  #> locale:
  #> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
  #> 
  #> attached base packages:
  #> [1] stats     graphics  grDevices utils     datasets  methods   base     
  #> 
  #> other attached packages:
  #> [1] BiocStyle_2.25.0
  #> 
  #> loaded via a namespace (and not attached):
  #>  [1] knitr_1.39          magrittr_2.0.3      R6_2.5.1            ragg_1.2.2          rlang_1.0.4        
  #>  [6] fastmap_1.1.0       highr_0.9           stringr_1.4.0       tools_4.2.0         xfun_0.31          
  #> [11] png_0.1-7           cli_3.3.0           jquerylib_0.1.4     systemfonts_1.0.4   htmltools_0.5.3    
  #> [16] yaml_2.3.5          digest_0.6.29       rprojroot_2.0.3     pkgdown_2.0.6       bookdown_0.27      
  #> [21] textshaping_0.3.6   BiocManager_1.30.18 purrr_0.3.4         sass_0.4.2          fs_1.5.2           
  #> [26] memoise_2.0.1       cachem_1.0.6        evaluate_0.15       rmarkdown_2.14      stringi_1.7.8      
  #> [31] compiler_4.2.0      bslib_0.4.0         desc_1.4.1          jsonlite_1.8.0
  ```
-->
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
