<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="igvR">
<title>Obtain and Display H3K27ac K562 track from the AnnotationHub • igvR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Obtain and Display H3K27ac K562 track from the AnnotationHub">
<meta property="og:description" content="igvR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">igvR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.17.18</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/v00.basicIntro.html">Introduction: a simple demo</a>
    <a class="dropdown-item" href="../articles/v01.stockGenome.html">Use a Stock Genome</a>
    <a class="dropdown-item" href="../articles/v02.customGenome.html">Use a Custom Genome</a>
    <a class="dropdown-item" href="../articles/v03.ctcfChIP.html">Explore CTCF ChIP-seq alignments, MACS2 narrowPeaks, Motif Matching and H3K4me3 methylation</a>
    <a class="dropdown-item" href="../articles/v04.pairedEnd.html">Paired-end Interaction Tracks</a>
    <a class="dropdown-item" href="../articles/v05.ucscTableBrowser.html">Obtain and Display H3K4Me3 K562 track from UCSC table browser</a>
    <a class="dropdown-item" href="../articles/v06.annotationHub.html">Obtain and Display H3K27ac K562 track from the AnnotationHub</a>
    <a class="dropdown-item" href="../articles/v07.gwas.html">GWAS Tracks</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="v06.annotationHub_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Obtain and Display H3K27ac K562 track from the AnnotationHub</h1>
                        <h4 data-toc-skip class="author">Paul Shannon</h4>
            
            <h4 data-toc-skip class="date">2022-07-25</h4>
      
      
      <div class="d-none name"><code>v06.annotationHub.Rmd</code></div>
    </div>

    
    
<style>
.main-container { width: 1000px; max-width:2800px;}
</style>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The Bioconductor AnnotationHub is a good source of genomic annotations of many different kinds.</p>
<p>H3K27ac is an epigenetic modification to the histone H3, an acetylation of the lysine residue at N-terminal position 27. H3K27ac is <a href="https://www.pnas.org/doi/10.1073/pnas.1016071107" class="external-link">associated with active enhancers</a>.</p>
<p>To the best of my knowledge, fetching data from the AnnotationHub does not support regions. The fetch is necessarily of the entire genomic resource - all chromosomes - and so may require time-consuming downloads. Subsetting by region takes place <strong>after</strong> the often time-consuming download.</p>
<p>Therefore, to run this vignette for the first time may take up to 20 minutes due to that download time.</p>
<p>Once downloaded, however, the resource is cached.</p>
</div>
<div class="section level2">
<h2 id="display-a-genomic-region-of-interest-in-igvr">Display a genomic region of interest in igvR<a class="anchor" aria-label="anchor" href="#display-a-genomic-region-of-interest-in-igvr"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://paul-shannon.github.io/igvR/" class="external-link">igvR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span></span>
<span><span class="va">igv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/igvR-class.html">igvR</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://paul-shannon.github.io/BrowserViz/reference/setBrowserWindowTitle.html" class="external-link">setBrowserWindowTitle</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"H3K27ac GATA2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/setGenome.html">setGenome</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"hg19"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/showGenomicRegion.html">showGenomicRegion</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"GATA2"</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span> <span class="fu"><a href="../reference/zoomOut.html">zoomOut</a></span><span class="op">(</span><span class="va">igv</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/annotationHub-01.png" width="95%"></p>
</div>
<div class="section level2">
<h2 id="query-the-annotationhub">Query the AnnotationHub<a class="anchor" aria-label="anchor" href="#query-the-annotationhub"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aHub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">query.terms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H3K27Ac"</span>, <span class="st">"k562"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">aHub</span>, <span class="va">query.terms</span><span class="op">)</span><span class="op">)</span>  <span class="co"># found 7</span></span>
<span><span class="va">h3k27ac.entries</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">aHub</span>, <span class="va">query.terms</span><span class="op">)</span></span></code></pre></div>
<p>These are the keys and titles of the available data</p>
<pre><code><span>             <span class="va">title</span></span>
<span>  <span class="va">AH23388</span> <span class="op">|</span> <span class="va">wgEncodeBroadHistoneK562H3k27acStdPk.broadPeak.gz</span></span>
<span>  <span class="va">AH29788</span> <span class="op">|</span> <span class="va">E123</span><span class="op">-</span><span class="va">H3K27ac.broadPeak.gz</span></span>
<span>  <span class="va">AH30836</span> <span class="op">|</span> <span class="va">E123</span><span class="op">-</span><span class="va">H3K27ac.narrowPeak.gz</span></span>
<span>  <span class="va">AH31772</span> <span class="op">|</span> <span class="va">E123</span><span class="op">-</span><span class="va">H3K27ac.gappedPeak.gz</span></span>
<span>  <span class="va">AH32958</span> <span class="op">|</span> <span class="va">E123</span><span class="op">-</span><span class="va">H3K27ac.fc.signal.bigwig</span></span>
<span>  <span class="va">AH33990</span> <span class="op">|</span> <span class="va">E123</span><span class="op">-</span><span class="va">H3K27ac.pval.signal.bigwig</span></span>
<span>  <span class="va">AH39539</span> <span class="op">|</span> <span class="va">E123</span><span class="op">-</span><span class="va">H3K27ac.imputed.pval.signal.bigwig</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="select-two-resources-boadpeaks-and-fc-bigwig">Select Two Resources: boadPeaks and fc bigwig<a class="anchor" aria-label="anchor" href="#select-two-resources-boadpeaks-and-fc-bigwig"></a>
</h2>
<p>If not in your cache, this step may take 20 minutes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x.broadPeak</span> <span class="op">&lt;-</span> <span class="va">aHub</span><span class="op">[[</span><span class="st">"AH23388"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">x.bigWig</span> <span class="op">&lt;-</span> <span class="va">aHub</span><span class="op">[[</span><span class="st">"AH32958"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>The two resources are different data types, requiring different processing to render as tracks in igvR</p>
<ul>
<li>
<strong>x.broadPeak</strong> is a GRanges object in memory</li>
<li>
<strong>x.bigWig</strong> is a bigwig file in your cache</li>
</ul>
</div>
<div class="section level2">
<h2 id="broadpeaks-subset-and-display">broadPeaks: subset and display<a class="anchor" aria-label="anchor" href="#broadpeaks-subset-and-display"></a>
</h2>
<p>The broadPeak data is a GRanges object already in memory. Subset to obtain only the 252 kb region in which we are interested.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">roi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGenomicRegion.html">getGenomicRegion</a></span><span class="op">(</span><span class="va">igv</span><span class="op">)</span></span>
<span><span class="va">gr.broadpeak</span> <span class="op">&lt;-</span> <span class="va">x.broadPeak</span><span class="op">[</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">x.broadpeak</span><span class="op">)</span><span class="op">==</span><span class="va">roi</span><span class="op">$</span><span class="va">chrom</span> <span class="op">&amp;</span></span>
<span>                            <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">x.broadpeak</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">roi</span><span class="op">$</span><span class="va">start</span> <span class="op">&amp;</span></span>
<span>                            <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">x.broadpeak</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">roi</span><span class="op">$</span><span class="va">end</span><span class="op">]</span></span></code></pre></div>
<p>igvR’s <strong>GrangesQuantitativeTrack</strong> must have only one numeric column in the GRanges metadata. That column is used as the magnitudes the track will display.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">gr.broadpeak</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co">#  "name"        "score"       "signalValue" "pValue"      "qValue"</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">gr.broadpeak</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gr.broadpeak</span><span class="op">$</span><span class="va">score</span></span>
<span><span class="va">track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRangesQuantitativeTrack-class.html">GRangesQuantitativeTrack</a></span><span class="op">(</span><span class="st">"h3k27ac bp"</span>, <span class="va">gr.broadpeak</span>, autoscale<span class="op">=</span><span class="cn">TRUE</span>, color<span class="op">=</span><span class="st">"brown"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/displayTrack.html">displayTrack</a></span><span class="op">(</span><span class="va">igv</span>, <span class="va">track</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="bigwig-subset-and-display">bigWig: subset and display<a class="anchor" aria-label="anchor" href="#bigwig-subset-and-display"></a>
</h2>
<p>We use the import function from the <strong>rtracklayer</strong> package to read in only a small portion of the large bigWig file. Note that, as read, there is only one numeric metadata colum, “score”, so no reduction of mcols is needed.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">file.bigWig</span> <span class="op">&lt;-</span> <span class="fu">resource</span><span class="op">(</span><span class="va">x.bigWig</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">gr.roi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">roi</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames<span class="op">=</span><span class="va">chrom</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="va">start</span>, <span class="va">end</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gr.bw</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">file.bigWig</span>, which<span class="op">=</span><span class="va">gr.roi</span>, format<span class="op">=</span><span class="st">"bigWig"</span><span class="op">)</span></span>
<span><span class="va">track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRangesQuantitativeTrack-class.html">GRangesQuantitativeTrack</a></span><span class="op">(</span><span class="st">"h3k27ac.bw"</span>, <span class="va">gr.bw</span>, autoscale<span class="op">=</span><span class="cn">TRUE</span>, color<span class="op">=</span><span class="st">"gray"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/displayTrack.html">displayTrack</a></span><span class="op">(</span><span class="va">igv</span>, <span class="va">track</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/annotationHub-02.png" width="95%"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur/Monterey 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] BiocStyle_2.25.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] knitr_1.39          magrittr_2.0.3      R6_2.5.1            ragg_1.2.2          rlang_1.0.4        </span></span>
<span><span class="co">#&gt;  [6] fastmap_1.1.0       highr_0.9           stringr_1.4.0       tools_4.2.0         xfun_0.31          </span></span>
<span><span class="co">#&gt; [11] cli_3.3.0           jquerylib_0.1.4     systemfonts_1.0.4   htmltools_0.5.3     yaml_2.3.5         </span></span>
<span><span class="co">#&gt; [16] digest_0.6.29       rprojroot_2.0.3     pkgdown_2.0.6       bookdown_0.27       textshaping_0.3.6  </span></span>
<span><span class="co">#&gt; [21] BiocManager_1.30.18 purrr_0.3.4         sass_0.4.2          fs_1.5.2            memoise_2.0.1      </span></span>
<span><span class="co">#&gt; [26] cachem_1.0.6        evaluate_0.15       rmarkdown_2.14      stringi_1.7.8       compiler_4.2.0     </span></span>
<span><span class="co">#&gt; [31] bslib_0.4.0         desc_1.4.1          jsonlite_1.8.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
