<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="igvR">
<title>GWAS Tracks • igvR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GWAS Tracks">
<meta property="og:description" content="igvR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">igvR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.17.21</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/v00.basicIntro.html">Introduction: a simple demo</a>
    <a class="dropdown-item" href="../articles/v01.stockGenome.html">Use a Stock Genome</a>
    <a class="dropdown-item" href="../articles/v02.customGenome.html">Use a Custom Genome</a>
    <a class="dropdown-item" href="../articles/v03.ctcfChIP.html">Explore CTCF ChIP-seq alignments, MACS2 narrowPeaks, Motif Matching and H3K4me3 methylation</a>
    <a class="dropdown-item" href="../articles/v04.pairedEnd.html">Paired-end Interaction Tracks</a>
    <a class="dropdown-item" href="../articles/v05.ucscTableBrowser.html">Obtain and Display H3K4Me3 K562 track from UCSC table browser</a>
    <a class="dropdown-item" href="../articles/v06.annotationHub.html">Obtain and Display H3K27ac K562 track from the AnnotationHub</a>
    <a class="dropdown-item" href="../articles/v07.gwas.html">GWAS Tracks</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="v07.gwas_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>GWAS Tracks</h1>
                        <h4 data-toc-skip class="author">Paul Shannon</h4>
            
            <h4 data-toc-skip class="date">2022-08-05</h4>
      
      
      <div class="d-none name"><code>v07.gwas.Rmd</code></div>
    </div>

    
    
<style>
.main-container { width: 1000px; max-width:2800px;}
</style>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Genome Wide Association Studies (GWAS) are often viewed in a “manhattan plot”. <a href="https://github.com/igvteam/igv.js/wiki/GWAS" class="external-link">igv.js</a> supports thesein two formats: a fixed bed format, and the traditional flexible, multi-column file format in which GWAS results are often summarized.</p>
<p>We support only the latter since its flexibility can be used with the five-column fixed bed format as well.</p>
<p>The data from GWAS studies is rather sparse, usually with no more than a few hundred variant hotspots which associate with the phenotype of interest. Another feature of these data is that the community is often most concerned with the statistical significance of the association. This is expressed as a pvalue. A minus-log10 transform of the pvalue controls the height of the rendered variant in the manhattant plot.</p>
<p>Support for this track type in igvR is new (summer 2022) and unpolished. Control of autoscaling and fixed scaling needs work.</p>
<p>There are two relevant classes, shown here with their constructors:</p>
<ol style="list-style-type: decimal">
<li>GWASTrack(“bellenguuez”, tbl.gwas, chrom.col=1, pos.col=2, pval.col=5, trackHeight=80)</li>
<li>GWASUrlTrack(“igv sample”, url, chrom.col=12, pos.col=13, pval.col=28)</li>
</ol>
<p>All the arguments are the same, except for the second one, which is either a local data.frame in memory, or a url pointing to a remote file served over http.</p>
</div>
<div class="section level2">
<h2 id="demostration-a-local-data-frame-with-5-columns">Demostration: a local data.frame with 5 columns<a class="anchor" aria-label="anchor" href="#demostration-a-local-data-frame-with-5-columns"></a>
</h2>
<p>Here we use a 5-column data.frame from a file hosted in igvR’s extdata, GWAS results from a 2022 Nature Genetics paper, <a href="https://www.nature.com/articles/s41588-022-01024-z" class="external-link">New insights into the genetic etiology of Alzheimer’s disease and related dementias</a>.</p>
<p>Its first few rows:</p>
<pre><code>  chrom     start       end        name    score
1  chr1 109345809 109345810 rs141749679 2.36e-04
2  chr1 207577222 207577223    rs679515 1.51e-14
3  chr2   9558881   9558882  rs72777026 4.14e-03
4  chr2  37304795  37304796  rs17020490 2.33e-04
5  chr2 105749598 105749599 rs143080277 6.29e-04
6  chr2 127135233 127135234   rs6733839 2.06e-30</code></pre>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://paul-shannon.github.io/igvR/" class="external-link">igvR</a></span><span class="op">)</span></span>
<span><span class="va">igv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/igvR-class.html">igvR</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://paul-shannon.github.io/BrowserViz/reference/setBrowserWindowTitle.html" class="external-link">setBrowserWindowTitle</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"AD GWAS"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/setGenome.html">setGenome</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"hg38"</span><span class="op">)</span></span>
<span><span class="va">tbl.gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"igvR"</span>, <span class="st">"extdata"</span>, <span class="st">"gwas"</span>, <span class="st">"bellenguez.bed"</span><span class="op">)</span>,</span>
<span>                       sep<span class="op">=</span><span class="st">"\t"</span>, as.is<span class="op">=</span><span class="cn">TRUE</span>, header<span class="op">=</span><span class="cn">TRUE</span>, nrow<span class="op">=</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GWASTrack-class.html">GWASTrack</a></span><span class="op">(</span><span class="st">"bellenguuez"</span>, <span class="va">tbl.gwas</span>, chrom.col<span class="op">=</span><span class="fl">1</span>, pos.col<span class="op">=</span><span class="fl">2</span>, pval.col<span class="op">=</span><span class="fl">5</span>, trackHeight<span class="op">=</span><span class="fl">80</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/displayTrack.html">displayTrack</a></span><span class="op">(</span><span class="va">igv</span>, <span class="va">track</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/gwas-01.png" width="95%"></p>
<p>Zoomed into the neighborhood of TREM2, a gene independently known to be associated with Alzheimer’s Disease.</p>
<p><img src="images/gwas-02.png" width="95%"></p>
</div>
<div class="section level2">
<h2 id="demonstration-a-remote-gwas-file-with-34-columns">Demonstration: a remote GWAS file with 34 columns<a class="anchor" aria-label="anchor" href="#demonstration-a-remote-gwas-file-with-34-columns"></a>
</h2>
<p>This file is more typical of that used in GWAS studies. These are the columns:</p>
<ol style="list-style-type: decimal">
<li>DATE ADDED TO CATALOG</li>
<li>PUBMEDID</li>
<li>FIRST AUTHOR</li>
<li>DATE</li>
<li>JOURNAL</li>
<li>LINK</li>
<li>STUDY</li>
<li>DISEASE/TRAIT</li>
<li>INITIAL SAMPLE SIZE</li>
<li>REPLICATION SAMPLE SIZE</li>
<li>REGION</li>
<li>CHR_ID</li>
<li>CHR_POS</li>
<li>REPORTED GENE(S)</li>
<li>MAPPED_GENE</li>
<li>UPSTREAM_GENE_ID</li>
<li>DOWNSTREAM_GENE_ID</li>
<li>SNP_GENE_IDS</li>
<li>UPSTREAM_GENE_DISTANCE</li>
<li>DOWNSTREAM_GENE_DISTANCE</li>
<li>STRONGEST SNP-RISK ALLELE</li>
<li>SNPS</li>
<li>MERGED</li>
<li>SNP_ID_CURRENT</li>
<li>CONTEXT</li>
<li>INTERGENIC</li>
<li>RISK ALLELE FREQUENCY</li>
<li>P-VALUE</li>
<li>PVALUE_MLOG</li>
<li>P-VALUE (TEXT)</li>
<li>OR or BETA</li>
<li>95% CI (TEXT)</li>
<li>PLATFORM [SNPS PASSING QC]</li>
<li>CNV</li>
</ol>
<p>Here we create the track and display it:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://s3.amazonaws.com/igv.org.demo/gwas_sample.tsv.gz"</span></span>
<span><span class="va">track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GWASUrlTrack-class.html">GWASUrlTrack</a></span><span class="op">(</span><span class="st">"igv sample"</span>, <span class="va">url</span>,chrom.col<span class="op">=</span><span class="fl">12</span>, pos.col<span class="op">=</span><span class="fl">13</span>, pval.col<span class="op">=</span><span class="fl">28</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/displayTrack.html">displayTrack</a></span><span class="op">(</span><span class="va">igv</span>, <span class="va">track</span><span class="op">)</span></span></code></pre></div>
<p>First, the whole genome view:</p>
<p><img src="images/gwas-03.png" width="95%"> Now, zooin into on chromsome 6, and click on any variant in the display (here we chose <strong>rs7767396</strong>) the full data from the multi-column gwas file is available.</p>
<p><img src="images/gwas-04.png" width="95%"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur/Monterey 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] BiocStyle_2.25.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] knitr_1.39          magrittr_2.0.3      R6_2.5.1            ragg_1.2.2          rlang_1.0.4        </span></span>
<span><span class="co">#&gt;  [6] fastmap_1.1.0       highr_0.9           stringr_1.4.0       tools_4.2.0         xfun_0.31          </span></span>
<span><span class="co">#&gt; [11] cli_3.3.0           jquerylib_0.1.4     systemfonts_1.0.4   htmltools_0.5.3     yaml_2.3.5         </span></span>
<span><span class="co">#&gt; [16] digest_0.6.29       rprojroot_2.0.3     pkgdown_2.0.6       bookdown_0.27       textshaping_0.3.6  </span></span>
<span><span class="co">#&gt; [21] BiocManager_1.30.18 purrr_0.3.4         sass_0.4.2          fs_1.5.2            memoise_2.0.1      </span></span>
<span><span class="co">#&gt; [26] cachem_1.0.6        evaluate_0.15       rmarkdown_2.14      stringi_1.7.8       compiler_4.2.0     </span></span>
<span><span class="co">#&gt; [31] bslib_0.4.0         desc_1.4.1          jsonlite_1.8.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
