<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="GFF3Track creates an IGV track for 9-column gene annotation tables"><title>Constructor for GFF3Track — GFF3Track-class • igvR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Constructor for GFF3Track — GFF3Track-class"><meta property="og:description" content="GFF3Track creates an IGV track for 9-column gene annotation tables"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">igvR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.17.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/alzheimersVariantsNearMEF2C.html">Explore VCF variants, GWAS snps, promoters and histone marks around the  MEF2C gene in Alzheimers Disease</a>
    <a class="dropdown-item" href="../articles/basicIntro.html">Introduction: a simple demo</a>
    <a class="dropdown-item" href="../articles/chooseStockOrCustomGenome.html">Choose aStock or Custom Genome</a>
    <a class="dropdown-item" href="../articles/ctcfChIP.html">Explore CTCF ChIP-seq alignments, MACS2 narrowPeaks, Motif Matching and H3K4me3 methylation</a>
    <a class="dropdown-item" href="../articles/ctcfChipSeq-old.html">Explore ChIP-seq alignments from a bam file, MACS2 narrowPeaks, conservation, H3K4me3 methylation and motif matching</a>
    <a class="dropdown-item" href="../articles/customGenome.html">Use a Custom Genome</a>
    <a class="dropdown-item" href="../articles/pairedEnd.html">Paired-end Interaction Tracks</a>
    <a class="dropdown-item" href="../articles/stockGenome.html">Use a Stock Genome</a>
    <a class="dropdown-item" href="../articles/ucscTableBrowser.html">Obtain and Display H3K4Me3 K562 track from UCSC table browser</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Constructor for GFF3Track</h1>
      
      <div class="d-none name"><code>GFF3Track-class.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>GFF3Track</code> creates an <code>IGV</code> track for 9-column gene annotation tables</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GFF3Track</span><span class="op">(</span></span>
<span>  <span class="va">trackName</span>,</span>
<span>  tbl.track <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  url <span class="op">=</span> <span class="cn">NA_character_</span>,</span>
<span>  indexURL <span class="op">=</span> <span class="cn">NA_character_</span>,</span>
<span>  trackColor <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  colorByAttribute <span class="op">=</span> <span class="cn">NA_character_</span>,</span>
<span>  colorTable <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">displayMode</span>,</span>
<span>  <span class="va">trackHeight</span>,</span>
<span>  <span class="va">visibilityWindow</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>trackName</dt>
<dd><p>A character string, used as track label by igv, we recommend unique names per track.</p></dd>


<dt>tbl.track</dt>
<dd><p>data.frame with 9 columns as defined at http://uswest.ensembl.org/info/website/upload/gff3.html</p></dd>


<dt>url</dt>
<dd><p>character the web location of a 9-column table, gzipped or not</p></dd>


<dt>indexURL</dt>
<dd><p>character the matching tabix index file</p></dd>


<dt>trackColor</dt>
<dd><p>character a recognized color name or RGB triple</p></dd>


<dt>colorByAttribute</dt>
<dd><p>a name from a column 9 attribute</p></dd>


<dt>colorTable</dt>
<dd><p>list which maps the colorByAttribute values to different colors</p></dd>


<dt>displayMode</dt>
<dd><p>"COLLAPSED", "SQUISHED" or "EXPANDED".  Spelling and case must be precise.</p></dd>


<dt>trackHeight</dt>
<dd><p>track height, typically in range 20 (for annotations) and up to 1000 (for large sample vcf files)</p></dd>


<dt>visibilityWindow</dt>
<dd><p>Maximum window size in base pairs for which indexed annotations or variants are displayed. Defaults: 1 MB for variants, whole chromosome for other track types.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A GFF3Track object</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Detailed description goes here</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">tbl.gff3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"igvR"</span>, <span class="st">"extdata"</span>, <span class="st">"GRCh38.94.NDUFS2.gff3"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                       sep<span class="op">=</span><span class="st">"\t"</span>, as.is<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbl.gff3</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqid"</span>, <span class="st">"source"</span>, <span class="st">"type"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"score"</span>, <span class="st">"strand"</span>,</span></span>
<span class="r-in"><span>                        <span class="st">"phase"</span>, <span class="st">"attributes"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"antisense"</span> <span class="op">=</span> <span class="st">"blueviolet"</span>,</span></span>
<span class="r-in"><span>               <span class="st">"protein_coding"</span> <span class="op">=</span> <span class="st">"blue"</span>,</span></span>
<span class="r-in"><span>               <span class="st">"retained_intron"</span> <span class="op">=</span> <span class="st">"rgb(0, 150, 150)"</span>,</span></span>
<span class="r-in"><span>               <span class="st">"processed_transcript"</span> <span class="op">=</span> <span class="st">"purple"</span>,</span></span>
<span class="r-in"><span>               <span class="st">"processed_pseudogene"</span> <span class="op">=</span> <span class="st">"#7fff00"</span>,</span></span>
<span class="r-in"><span>               <span class="st">"unprocessed_pseudogene"</span> <span class="op">=</span> <span class="st">"#d2691e"</span>,</span></span>
<span class="r-in"><span>               <span class="st">"default"</span> <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="va">track</span> <span class="op">&lt;-</span> <span class="fu">GFF3Track</span><span class="op">(</span><span class="st">"dataframe gff3"</span>, <span class="va">tbl.gff3</span>, colorByAttribute<span class="op">=</span><span class="st">"biotype"</span>, colorTable<span class="op">=</span><span class="va">colors</span>,</span></span>
<span class="r-in"><span>                   url<span class="op">=</span><span class="cn">NA_character_</span>, indexURL<span class="op">=</span><span class="cn">NA_character_</span>, displayMode<span class="op">=</span><span class="st">"EXPANDED"</span>, trackHeight<span class="op">=</span><span class="fl">200</span>,</span></span>
<span class="r-in"><span>                   visibilityWindow<span class="op">=</span><span class="fl">100000</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

